kotlinx-coroutines-core/api/kotlinx-coroutines-core.api

```diff
@@ -858,19 +863,22 @@ public final class kotlinx/coroutines/channels/ProducerScope$DefaultImpls {
 }
 
 public abstract interface class kotlinx/coroutines/channels/ReceiveChannel {
-	public abstract synthetic fun cancel ()V
+	public synthetic fun cancel ()V
 	public abstract synthetic fun cancel (Ljava/lang/Throwable;)Z
 	public abstract fun cancel (Ljava/util/concurrent/CancellationException;)V
+	public static synthetic fun cancel$default (Lkotlinx/coroutines/channels/ReceiveChannel;Ljava/lang/Throwable;ILjava/lang/Object;)Z
+	public static synthetic fun cancel$default (Lkotlinx/coroutines/channels/ReceiveChannel;Ljava/util/concurrent/CancellationException;ILjava/lang/Object;)V
 	public abstract fun getOnReceive ()Lkotlinx/coroutines/selects/SelectClause1;
 	public abstract fun getOnReceiveCatching ()Lkotlinx/coroutines/selects/SelectClause1;
-	public abstract fun getOnReceiveOrNull ()Lkotlinx/coroutines/selects/SelectClause1;
+	public fun getOnReceiveOrNull ()Lkotlinx/coroutines/selects/SelectClause1;
 	public abstract fun isClosedForReceive ()Z
 	public abstract fun isEmpty ()Z
 	public abstract fun iterator ()Lkotlinx/coroutines/channels/ChannelIterator;
-	public abstract fun poll ()Ljava/lang/Object;
+	public fun poll ()Ljava/lang/Object;
 	public abstract fun receive (Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
 	public abstract fun receiveCatching-JP2dKIU (Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
-	public abstract fun receiveOrNull (Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
+	public fun receiveOrNull (Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
+	public static synthetic fun receiveOrNull$suspendImpl (Lkotlinx/coroutines/channels/ReceiveChannel;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
```

Comment:
Are these `$suspendImpl` methods now also part of the public ABI? I've seen them too when trying out `all-compatibility`/`all` mode and was wondering what these methods are used for. It seems to have something to do with [KT-49106](https://youtrack.jetbrains.com/issue/KT-49106) fixed with https://github.com/JetBrains/kotlin/pull/4608. But couldn't the visibility be private? Methods in interfaces are allowed to be private since class file version 52.0 (Java 8).

Reply:
OK, I see why the `$suspendImpl` methods can't be private, they are used from the state machine class of the suspend function (if it has one). So if that's the only use, it should probably be filtered out by [BCV](https://github.com/Kotlin/binary-compatibility-validator), right?

Reply:
This example doesn't use `receiveOrNull$suspendImpl` for the `super` call, it uses `receiveOrNull` with `invokespecial`:

```
Classfile .../bar/Foo.class
  ...
public interface bar.Foo extends kotlinx.coroutines.channels.ReceiveChannel<java.lang.Integer>
  minor version: 0
  major version: 52
  flags: (0x0601) ACC_PUBLIC, ACC_INTERFACE, ACC_ABSTRACT
  this_class: #2                          // bar/Foo
  super_class: #5                         // java/lang/Object
  interfaces: 1, fields: 0, methods: 4, attributes: 4
Constant pool:
   ...
{
  public default java.lang.Object foo(kotlin.coroutines.Continuation<? super kotlin.Unit>);
    descriptor: (Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
    flags: (0x0001) ACC_PUBLIC
    Code:
      stack=2, locals=2, args_size=2
         0: aload_0
         1: aload_1
         2: invokestatic  #16                 // InterfaceMethod foo$suspendImpl:(Lbar/Foo;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
         5: areturn
      ...
    ...

  public static java.lang.Object foo$suspendImpl(bar.Foo, kotlin.coroutines.Continuation<? super kotlin.Unit>);
    descriptor: (Lbar/Foo;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
    flags: (0x1009) ACC_PUBLIC, ACC_STATIC, ACC_SYNTHETIC
    Code:
      stack=3, locals=2, args_size=2
         0: aload_0
         1: aload_1
         2: invokespecial #24                 // InterfaceMethod kotlinx/coroutines/channels/ReceiveChannel.receiveOrNull:(Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
         5: dup
         6: invokestatic  #30                 // Method kotlin/coroutines/intrinsics/IntrinsicsKt.getCOROUTINE_SUSPENDED:()Ljava/lang/Object;
         9: if_acmpne     13
        12: areturn
        13: pop
        14: getstatic     #36                 // Field kotlin/Unit.INSTANCE:Lkotlin/Unit;
        17: areturn
      ...
    ...

  ...
}
InnerClasses:
  public static final #59= #58 of #2;     // DefaultImpls=class bar/Foo$DefaultImpls of class bar/Foo
...
```

Reply:
```
interface Foo : ReceiveChannel<Int> {
    @Suppress("DEPRECATION_ERROR")
    override suspend fun receiveOrNull(): Int? {
        return super.receiveOrNull()
    }
}
```

```
  public default java.lang.Object receiveOrNull(kotlin.coroutines.Continuation<? super java.lang.Integer>);
    descriptor: (Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
    flags: (0x0001) ACC_PUBLIC
    Code:
      stack=2, locals=2, args_size=2
         0: aload_0
         1: aload_1
         2: invokestatic  #16                 // InterfaceMethod receiveOrNull$suspendImpl:(Lorg/example/Foo;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
         5: areturn
```

Reply:
@fzhinkin this is probably referring to the `receiveOrNull$suspendImpl` static method in the `Foo` interface itself (given that its descriptor includes `Lorg/example/Foo;`), take a look at the `InterfaceMethodref` constant pool entry at index 16. For me it looks like this:
```
Constant pool:
   #1 = Utf8               org/example/Foo
   #2 = Class              #1             // org/example/Foo
  ...
  #13 = Utf8               receiveOrNull$suspendImpl
  #14 = Utf8               (Lorg/example/Foo;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
  #15 = NameAndType        #13:#14        // receiveOrNull$suspendImpl:(Lorg/example/Foo;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
  #16 = InterfaceMethodref #2.#15         // org/example/Foo.receiveOrNull$suspendImpl:(Lorg/example/Foo;Lkotlin/coroutines/Continuation;)Ljava/lang/Object
  ...
```

The `receiveOrNull$suspendImpl` static method in `Foo` still calls `ReceiveChannel.receiveOrNull` via `invokespecial` like above:
```
  public static java.lang.Object receiveOrNull$suspendImpl(org.example.Foo, kotlin.coroutines.Continuation<? super java.lang.Integer>);
    descriptor: (Lorg/example/Foo;Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
    flags: (0x1009) ACC_PUBLIC, ACC_STATIC, ACC_SYNTHETIC
    Code:
      stack=2, locals=2, args_size=2
         0: aload_0
         1: aload_1
         2: invokespecial #23                 // InterfaceMethod kotlinx/coroutines/channels/ReceiveChannel.receiveOrNull:(Lkotlin/coroutines/Continuation;)Ljava/lang/Object;
         5: areturn
      ...
```

Reply:
@lukellmann nice catch! Interface's `$suspendImpl` methods seems only be public to conform with JVM spec. I'll double check, and if it's indeed the case, corresponding filtration should be added to BCV.

Reply:
https://github.com/Kotlin/binary-compatibility-validator/issues/270